cirq/ops/controlled_gate.py

```diff
@@ -12,52 +12,102 @@
 # See the License for the specific language governing permissions and
 # limitations under the License.
 
-from typing import Any, Union
+from typing import Any, List, Union
 
 import numpy as np
 
 import cirq
 from cirq import linalg, protocols, value
-from cirq.ops import raw_types, controlled_operation as cop
+from cirq.ops import raw_types, controlled_operation as cop, gate_operation
 from cirq.type_workarounds import NotImplementedType
 
 
 @value.value_equality
 class ControlledGate(raw_types.Gate):
     """Augments existing gates with a control qubit."""
 
-    def __init__(self, sub_gate: raw_types.Gate) -> None:
+    def __init__(self, sub_gate: raw_types.Gate,
+                 control_qubits: List[raw_types.Qid] = None,
+                 num_controls: int = 1) -> None:
```

Comment:
Currently, if a user calls this like `ControlledGate(g, control_qubits=[a, b])` they will get a confusing exception about the `num_controls` argument.

Have `num_controls` default to `None`. If both `control_qubits` and `num_controls` are specified (i.e. not `None`), raise an exception about only specifying one of them. If neither is set, default to 1 unspecified control qubit. If just `num_controls` is set, default `control_qubits` to `[None]*num_controls`.

Reply:
Agreed. But instead of always raising an exception when control_qubits and num_controls are specified (i.e. not None) we can handle cases where num_controls>len(control_qubits) by considering the others as yet-to-be-specified